rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/RemoteFetchProcessor.java

```diff
@@ -0,0 +1,211 @@
+
+  public RemoteFetchProcessor(BookKeeper bookKeeper, Configuration conf)
+  {
+    this.bookKeeper = bookKeeper;
+    this.conf = conf;
+    this.processQueue = new ConcurrentLinkedQueue<FetchRequest>();
+
+    this.diskReadBufferSize = CacheConfig.getDiskReadBufferSizeDefault(conf);
+    this.processThreadInitalDelay = CacheConfig.getProcessThreadInitialDelayInMs(conf);
+    this.processThreadInterval = CacheConfig.getProcessThreadIntervalInMs(conf);
+    this.requestProcessDelay = CacheConfig.getRemoteFetchProcessIntervalInMS(conf);
+    int numThreads = CacheConfig.getRemoteFetchNumThreads(conf);
+
+    ThreadPoolExecutor executor = (ThreadPoolExecutor) Executors.newFixedThreadPool(numThreads);
+    processService = MoreExecutors.getExitingExecutorService(executor);
+  }
+
+  public void addToProcessQueue(String remotePath, long offset, int length, long fileSize, long lastModified)
+  {
+    long requestedTime = System.currentTimeMillis();
+    FetchRequest request = new FetchRequest(remotePath, offset, length, fileSize, lastModified, requestedTime);
+    processQueue.add(request);
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception
+  {
+    long currentTime = System.currentTimeMillis();
+
+    processRequest(currentTime);
+  }
+
+  private void processRequest(long currentTime) throws IOException, InterruptedException, ExecutionException
+  {
+    ConcurrentMap<String, FileMetadataRequest> fetchRequestMap = new ConcurrentHashMap<String, FileMetadataRequest>();
+    ConcurrentMap<String, RangeSet<Long>> rangeMap = new ConcurrentHashMap<String, RangeSet<Long>>();
+
+    // Till the queue is not empty or there are no more requests which came in before the configured delay time
+    // we are going to collect the requests and process them
+    while (!processQueue.isEmpty()) {
+      FetchRequest request = processQueue.peek();
+      if (currentTime - request.getRequestedTime() < this.requestProcessDelay) {
+        break;
+      }
+
+      if (!rangeMap.containsKey(request.getRemotePath())) {
+        RangeSet<Long> rangeSet = TreeRangeSet.create();
+        rangeMap.putIfAbsent(request.getRemotePath(), rangeSet);
+        fetchRequestMap.putIfAbsent(request.getRemotePath(), new FileMetadataRequest(request.getRemotePath(),
+            request.getFileSize(), request.getLastModified()));
+      }
+      rangeMap.get(request.getRemotePath()).add(Range.open(request.getOffset(),
+          request.getOffset() + request.getLength()));
+      processQueue.remove();
+    }
+
+    processRemoteFetchRequest(fetchRequestMap, rangeMap);
+
+    // After every iteration we are clearing the maps
+    rangeMap.clear();
+    fetchRequestMap.clear();
+  }
+
+  @Override
+  protected Scheduler scheduler()
+  {
+    return Scheduler.newFixedDelaySchedule(processThreadInitalDelay, processThreadInterval, TimeUnit.MILLISECONDS);
+  }
+
+  private void processRemoteFetchRequest(ConcurrentMap<String, FileMetadataRequest> fetchRequestMap,
```

Comment:
Can we rename fetchRequestMap? Probably to fileMetadataMap, current name threw me off